modules/webgl/src/adapter/resources/webgl-transform-feedback.ts

```diff
@@ -0,0 +1,253 @@
+        const range = this._getBufferRange(bufferOrRange as WEBGLBuffer | BufferRange);
+        if (location < 0) {
+          this.unusedBuffers[locationOrName] = range.buffer;
+          log.warn(`${this.id} unused varying buffer ${locationOrName}`)();
+          return this;
+        }
+        this.buffers[location] = range;
+      }
+    } finally {
+      this.gl2.bindTransformFeedback(GL.TRANSFORM_FEEDBACK, null);
+    }
+
+    // Need to avoid chrome bug where buffer that is already bound to a different target
+    // cannot be bound to 'TRANSFORM_FEEDBACK_BUFFER' target.
+    if (!this.bindOnUse) {
+      this._bindBuffers();
+    }
+
+    Object.seal(this);
+  }
+
+  static override isSupported(device: Device | WebGLRenderingContext): boolean {
+    const webglDevice = WebGLDevice.attach(device);
+    return isWebGL2(webglDevice.gl);
+  }
+
+  override destroy(): void {
+    this.gl2.deleteTransformFeedback(this.handle);
+    super.destroy();
+  }
+
+  begin(primitiveMode = GL.POINTS): this {
+    this.gl2.bindTransformFeedback(GL.TRANSFORM_FEEDBACK, this.handle);
+    if (this.bindOnUse) {
+      this._bindBuffers();
+    }
+    this.gl2.beginTransformFeedback(primitiveMode);
+    return this;
+  }
+
+  end(): this {
+    this.gl2.endTransformFeedback();
+    if (!this.bindOnUse) {
+      this._unbindBuffers();
+    }
+    this.gl2.bindTransformFeedback(GL.TRANSFORM_FEEDBACK, null);
+    return this;
+  }
+
+  /**
+   * Need to avoid chrome bug where buffer that is already bound to a different target
+   * cannot be bound to 'TRANSFORM_FEEDBACK_BUFFER' target.
+   */
+  _bindBuffers(): void {
+    for (const bufferIndex in this.buffers) {
+      const {buffer, byteLength, byteOffset} = this._getBufferRange(this.buffers[bufferIndex]);
+      const handle = buffer && buffer.handle;
+      if (!handle || byteLength === undefined) {
+        this.gl2.bindBufferBase(GL.TRANSFORM_FEEDBACK_BUFFER, Number(bufferIndex), handle);
+      } else {
+        this.gl2.bindBufferRange(
+          GL.TRANSFORM_FEEDBACK_BUFFER,
+          Number(bufferIndex),
+          handle,
+          byteOffset,
+          byteLength
+        );
+      }
+    }
+  }
+
+  _unbindBuffers(): void {
+    for (const bufferIndex in this.buffers) {
+      this.gl2.bindBufferBase(GL.TRANSFORM_FEEDBACK_BUFFER, Number(bufferIndex), null);
+    }
+  }
+
+  _bindBuffer(
+    index: number,
+    buffer: WEBGLBuffer,
+    byteOffset: number = 0,
+    byteLength?: number
+  ): this {
+    const handle = buffer && buffer.handle;
+    if (!handle || byteLength === undefined) {
+      // @ts-expect-error
+      this.gl.bindBufferBase(GL.TRANSFORM_FEEDBACK_BUFFER, index, handle);
+    } else {
+      // @ts-expect-error
+      this.gl.bindBufferRange(GL.TRANSFORM_FEEDBACK_BUFFER, index, handle, byteOffset, byteLength);
+    }
+    return this;
+  }
+
+  // SUBCLASS
+
+  initialize(props?: TransformFeedbackProps): this {
```

Comment:
For most objects v9 longer supports reinitialization so this can be moved to the constructor